test/common/firefox-cdp-driver.js

```diff
@@ -0,0 +1,425 @@
+#!/usr/bin/env node
+
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2019 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/* firefox-cdp-driver -- A command-line JSON input/output wrapper around
+ * chrome-remote-interface (Chrome Debug Protocol).
+ * See https://chromedevtools.github.io/devtools-protocol/
+ * This needs support for protocol version 1.3.
+ *
+ * Set $TEST_CDP_DEBUG environment variable to enable additional
+ * frame/execution context debugging.
+ */
+
+const CDP = require('chrome-remote-interface');
+
+var enable_debug = false;
+var the_client = null;
+var last_frame_name = "";
+
+function debug(msg) {
+    if (enable_debug)
+        process.stderr.write("CDP: " + msg + "\n");
+}
+
+/**
+ * Format response to the client
+ */
+
+function fatal() {
+    console.error.apply(console.error, arguments);
+    process.exit(1);
+}
```

Comment:
There's a lot of duplicated code here. What's the next step for these two files? Do you want to work on unifying them again, or are there enough differences that these need to be two separate drivers for a while? In the latter case, could these move out the commonalities into a library?

I only coarsely reviewed this file, as it's mostly a question of "does it work?" and that smells like it's going to get more cleanup in the near future. :grin:

Reply:
Well, in a ideal world we would use only one file (the chromium one) and I hope that in some time we can do that. However right now this firefox driver contains a lot of hacks and workaround since the firefox API is very limited.
I don't think we should invest much time to unify them now, but as firefox implements more and more API I'll be happy to  move them closer to each other.  I would also like to point out that firefox CDP is becoming active and things are moving forward so I really hope that the functionality we need could be implemented in some reasonable time frame.

> could these move out the commonalities into a library?

I guess? There are some common functions but they all interact with `process` - which I am not sure if it can be provided by library, but I guess why not? I mostly just did not want to touch the chromium driver. But if you think it is worth it I'll go for it

Reply:
Not right now. The chromium driver hasn't changed in some time. We should do that if/when we need to change it, but chances are that some firefox hacks went away until then. So indeed it seems sensible for now to wait.